src/booster/builder.rs

```diff
@@ -0,0 +1,96 @@
+use serde_json::Value;
+
+use booster::Booster;
+use dataset::DataSet;
+use Error;
+use {InputMatrix, OutputVec};
+
+// types for training set
+#[derive(Clone)]
+pub struct TrainDataAdded(DataSet); // this should not implement default, so it can safely be used for construction
+#[derive(Default, Clone)]
+pub struct TrainDataNotAdded;
+
+// types for params
+#[derive(Clone)]
+pub struct ParamsAdded(Value); // this should not implement default, so it can safely be used for construction
+#[derive(Default, Clone)]
+pub struct ParamsNotAdded;
+
+/// Builder for the Booster.
+///
+/// Uses TypeState Pattern to make sure that Training Data is added
+/// so that Validation can be synced properly and params are present for training.
+#[derive(Default, Clone)]
+pub struct BoosterBuilder<T: Clone, P: Clone> {
+    train_data: T,
+    val_data: Vec<DataSet>,
+    params: P, // after #3 should this be a struct
+}
```

Comment:
It's a neat pattern, and it looks like Rustdoc doesn't have issues with it either.